File: python/luhn/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = [c for c in card_num if c != ' ']

    def valid(self):
        if any([not c.isdigit() for c in self.card_num]) or len(self.card_num) < 2:
            return False

        vals = [int(c) for c in self.card_num]
        for i in range(len(vals)-2, -1, -2):
            vals[i] *= 2
            if vals[i] > 9:
                vals[i] -= 9

        return sum(vals) % 10 == 0

File: csharp/luhn/Luhn.cs
using System.Collections.Generic;
using System.Linq;

public static class Luhn
{
    public static bool IsValid(string number)
    {
        string cleaned = number.Replace(" ", "");
        if (cleaned.Length < 2) return false;
        
        var digits = new List<int>();
        foreach (char c in cleaned)
        {
            if (!char.IsDigit(c)) return false;
            digits.Add((int)char.GetNumericValue(c));
        }
        digits.Reverse();
        for (int i = 1; i < digits.Count; i += 2)
        {
            digits[i] *= 2;
            if (digits[i] > 9) digits[i] -= 9;
        }
        return digits.Sum() % 10 == 0;
    }
}
